contracts/xasset/src/token.rs

```diff
@@ -157,35 +185,62 @@ impl Token {
         decimals: u32,
         annual_interest_rate: u32,
     ) {
-        let mut token = Token::default();
-        token.xlm_sac.set(&xlm_sac);
-        token.xlm_contract.set(&xlm_contract);
-        token.asset_contract.set(&asset_contract);
-        token.pegged_asset.set(&pegged_asset);
-        token.min_collat_ratio.set(&min_collat_ratio);
-        token.name.set(&name);
-        token.symbol.set(&symbol);
-        token.decimals.set(&decimals);
-        token.total_xasset.set(&0);
-        token.total_collateral.set(&0);
-        token.product_constant.set(&PRODUCT_CONSTANT);
-        token.compounded_constant.set(&0);
-        token.epoch.set(&0);
-        token.fees_collected.set(&0);
-        token.deposit_fee.set(&DEPOSIT_FEE);
-        token.stake_fee.set(&STAKE_FEE);
-        token.unstake_return.set(&UNSTAKE_RETURN);
-        token.interest_collected.set(&0);
-        token.interest_rate.set(&annual_interest_rate);
+        Self::set_admin(env, &admin);
+        let token = TokenStorage {
+            name,
+            symbol,
+            decimals,
+            xlm_sac,
+            xlm_contract,
+            asset_contract,
+            pegged_asset,
+            min_collat_ratio,
+            total_xasset: 0,
+            total_collateral: 0,
+            product_constant: PRODUCT_CONSTANT,
+            compounded_constant: 0,
+            epoch: 0,
+            fees_collected: 0,
+            deposit_fee: DEPOSIT_FEE,
+            stake_fee: STAKE_FEE,
+            unstake_return: UNSTAKE_RETURN,
+            interest_rate: annual_interest_rate,
+            interest_collected: 0,
+        };
+        TokenStorage::set_state(env, &token);
+    }
+
+    /// Upgrade the contract to new wasm. Admin-only.
+    pub fn upgrade(env: &Env, new_wasm_hash: BytesN<32>) {
+        Self::require_admin(env);
+        env.deployer().update_current_contract_wasm(new_wasm_hash);
     }
-}
 
-impl IsSep41 for Token {
```

Comment:
Instead of losing the trait impl you can add `#[contractimpl]` above it. This way you can still get the separation of traits without just comibing all into `impl` block. The one thing to know is that without `contracttait`s you need to copy the doc strings here.